### ISSUE TITLE
BusySpot error caused by timezone

### ISSUE THREAD

**Issue by hrenard** [SergeyMosin/Appointments#478]
Hi  !
Firstly thanks for your work !

I seem to be hitting an error because of some timezone discrepancy. 

## Facts

- External mode is used.
- The source is a shared calendar.
- Our time zone is Europe/Paris (+1).
- A slot is booked for 9:30 to 12:30.

If I try to book an available slot at 13h, it fails. But a slot starting at 13:30, would work.

## Hypotheses

The gap between the two slot would then be equivalent to the TZ offset which made me suspicious. After gathering some info, I manually modified the timezone of the booked event, and retried to book the 13h slot and it worked.

## Log

```
BusySpot/error detected in setAttendee func, mode: 1, trc_code: 1
```

## vCards

Source for the targeted slot:
```
BEGIN:VCALENDAR
VERSION:2.0
CALSCALE:GREGORIAN
PRODID:-//Mozilla.org/NONSGML Mozilla Calendar V1.1//EN
BEGIN:VTIMEZONE
TZID:Europe/Paris
BEGIN:DAYLIGHT
TZOFFSETFROM:+0100
TZOFFSETTO:+0200
TZNAME:CEST
DTSTART:19700329T020000
RRULE:FREQ=YEARLY;BYMONTH=3;BYDAY=-1SU
END:DAYLIGHT
BEGIN:STANDARD
TZOFFSETFROM:+0200
TZOFFSETTO:+0100
TZNAME:CET
DTSTART:19701025T030000
RRULE:FREQ=YEARLY;BYMONTH=10;BYDAY=-1SU
END:STANDARD
END:VTIMEZONE
BEGIN:VEVENT
CREATED:20231010T152140Z
LAST-MODIFIED:20240221T080305Z
DTSTAMP:20231010T154120Z
UID:d9bdca78-2ab5-4d85-a4ce-06e3acdae0c5
SUMMARY:_Après-midi
RRULE:FREQ=DAILY
DTSTART;TZID=Europe/Paris:20240214T130000
DTEND;TZID=Europe/Paris:20240214T180000
TRANSP:TRANSPARENT
X-MOZ-GENERATION:7
SEQUENCE:14
END:VEVENT
BEGIN:VEVENT
CREATED:20231010T152226Z
LAST-MODIFIED:20231010T154120Z
DTSTAMP:20231010T154120Z
UID:d9bdca78-2ab5-4d85-a4ce-06e3acdae0c5
SUMMARY:Après-midi
RECURRENCE-ID;TZID=Europe/Paris:20231009T130000
DTSTART;TZID=Europe/Paris:20231009T180000
DTEND;TZID=Europe/Paris:20231009T230000
TRANSP:TRANSPARENT
X-MOZ-GENERATION:7
SEQUENCE:1
END:VEVENT
END:VCALENDAR
```

Booked slot:
```
BEGIN:VCALENDAR
CALSCALE:GREGORIAN
VERSION:2.0
PRODID:-//IDN nextcloud.com//***//EN
BEGIN:VEVENT
SUMMARY:***
DTSTART:20240318T093000
DTEND:20240318T123000
DTSTAMP:20240109T154213Z
UID:4ae32c28-1f1e-48a6-a9e3-a5ccda584a58
CATEGORIES:Appointment
DESCRIPTION:***
LOCATION:***
STATUS:CONFIRMED
TRANSP:OPAQUE
X-APPT-DSR:***
X-TZX:UEurope/Paris
END:VEVENT
END:VCALENDAR
```

Modified/fixed booked slot:
```
BEGIN:VCALENDAR
CALSCALE:GREGORIAN
VERSION:2.0
PRODID:-//IDN nextcloud.com//***//EN
BEGIN:VEVENT
SUMMARY:***
DTSTART;TZID=Europe/Paris:20240318T093000
DTEND;TZID=Europe/Paris:20240318T123000
DTSTAMP:20240227T160534Z
UID:4ae32c28-1f1e-48a6-a9e3-a5ccda584a58
CATEGORIES:Appointment
DESCRIPTION:***
LOCATION:***
STATUS:CONFIRMED
TRANSP:OPAQUE
X-APPT-DSR:***
X-TZX:UEurope/Paris
SEQUENCE:1
LAST-MODIFIED:20240227T160534Z
END:VEVENT
BEGIN:VTIMEZONE
TZID:Europe/Paris
BEGIN:DAYLIGHT
TZOFFSETFROM:+0100
TZOFFSETTO:+0200
TZNAME:CEST
DTSTART:19700329T020000
RRULE:FREQ=YEARLY;BYMONTH=3;BYDAY=-1SU
END:DAYLIGHT
BEGIN:STANDARD
TZOFFSETFROM:+0200
TZOFFSETTO:+0100
TZNAME:CET
DTSTART:19701025T030000
RRULE:FREQ=YEARLY;BYMONTH=10;BYDAY=-1SU
END:STANDARD
END:VTIMEZONE
END:VCALENDAR
```

**Comment by SergeyMosin** [comment]
Thank you for the report, while I try to replicate this, can you please check how the Appointments App 'sees' the timezones of **both** the `Source` and the `Destination` calendars.

This can be done in `Appointments > Settings > Advanced Settings > Debugging > Get raw calendar data`.

Example output (timezone is in the second array)
```
array (
  'name' => 'Appt',
  'id' => '92',
  'isReadOnly' => '0',
  'isSubscription' => '0',
)
array (
  'id' => '92',
  'displayName' => 'Appt',
  'color' => '#93B27B',
  'uri' => 'appt',
  'timezone' => 'BEGIN:VCALENDAR
PRODID:-//IDN nextcloud.com//Calendar app 4.3.3//EN
CALSCALE:GREGORIAN
VERSION:2.0
BEGIN:VTIMEZONE
TZID:America/New_York
BEGIN:DAYLIGHT
...
```

**Comment by hrenard** [comment]
Yes, I sent it to you via email because there is private data.

**Comment by SergeyMosin** [comment]
Thank you for the debugging data. I think see what is happening. Some appointments/events ( created/added by 3rd party apps/scripts) use `floating`/`local time` format ( https://www.kanzaki.com/docs/ical/dateTime.html ) .

Example:
```
DTSTART:20240318T093000
DTEND:20240318T123000
```
which specs as

> Date-time values of this type are said to be "floating" and are not bound to any time zone in particular. They are used to represent the same hour, minute, and second value regardless of which time zone is currently being observed...

In such case the availability check is supposed  to fallback to the timezone of the `Destination` calendar. However, timezone info is lost during timestamp conversion and the system fallsback to UTC. The fix should be available in `v2.0.1` or `v2.0.2`

Bug location:  https://github.com/SergeyMosin/Appointments/blob/5474bae22f175f161080d584c3bab5f94f25b273/lib/Backend/BCSabreImpl.php#L139-L141
> The $timezone parameter and the current timezone are ignored when the $datetime parameter either is a UNIX timestamp (e.g. @946684800) or specifies a timezone (e.g. 2010-01-28T15:00:00+02:00).

**Comment by hrenard** [comment]
Great news ! Thanks for this fast debugging !

**Comment by JocelynDelalande** [comment]
Hi there :) @SergeyMosin any update on this ? I don't see the patch mentioned in changelog of 2.0.1/2/3

**Comment by SergeyMosin** [comment]
Fixed in v2.0.7
